### ISSUE TITLE
Spurious error "The requirement ... is added only by the Drop impl"

### ISSUE THREAD

**Issue by canndrew** [rust-lang/rust#34426]
It seems like it's not possible to implement `Drop` for types that have higher ranked trait bounds on their type arguments.

``` rust
trait WithLifetime<'a> {
}

struct Foo<A: for<'a> WithLifetime<'a>> {
    a: A,
}

impl<A: for<'a> WithLifetime<'a>> Drop for Foo<A> {
    fn drop(&mut self) {}
}
```

This gives the error _"The requirement `for<'a> A: WithLifetime<'a>` is added only by the Drop impl"_ which is obviously untrue. Removing that bound from the `Drop` impl gives the error _"the trait bound `for<'a> A: WithLifetime<'a>` is not satisfied"_ as it should.

**Comment by durka** [comment]
As an aside, the error message rephrased the bound to `for<'a> A: WithLifetime<'a>` which is illegal syntax.

**Comment by Osspial** [comment]
I'd be willing to try and tackle this issue! Is there anyone that would be able to do any mentoring? The attempts I've made to figure out the proper way of fixing this have resulted in me getting lost in the compiler's code, and I'm running out of stamina going at it alone.
